File: yatc-connections/src/test/java/fr/alexandreroman/yatc/connections/ApplicationTests.java
/*
 * Copyright (c) 2019 Pivotal Software, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.alexandreroman.yatc.connections;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.test.context.jdbc.SqlGroup;
import org.springframework.test.context.junit4.SpringRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.BDDMockito.given;

@ActiveProfiles("test")
@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class ApplicationTests {
    @LocalServerPort
    private int webPort;
    @Autowired
    private TestRestTemplate restTemplate;

    @MockBean
    private UsersClientService usersClientService;

    @Test
    public void contextLoads() {
    }

    private final String url(String path) {
        return "http://localhost:" + webPort + path;
    }

    @Test
    public void testGetFollowersUnknown() {
        given(usersClientService.userExists(anyString())).willReturn(false);

        final ResponseEntity<FollowersResponse> resp = restTemplate.getForEntity(url("/api/v1/connections/unknown"), FollowersResponse.class);
        assertThat(resp.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);
    }

    @Test
    public void testAddConnectionFromScratch() {
        given(usersClientService.userExists(anyString())).willReturn(true);

        final ResponseEntity<FollowersResponse> resp = restTemplate.postForEntity(url("/api/v1/connections/newuser/randomfollower"), null, FollowersResponse.class);
        assertThat(resp.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(resp.getBody().getFollowers().size()).isEqualTo(1);
        assertThat(resp.getBody().getFollowers().get(0)).isEqualTo("randomfollower");

        final ResponseEntity<FollowersResponse> resp2 = restTemplate.postForEntity(url("/api/v1/connections/newuser/randomfollower2"), null, FollowersResponse.class);
        assertThat(resp2.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(resp2.getBody().getFollowers().size()).isEqualTo(2);
        assertThat(resp2.getBody().getFollowers()).contains("randomfollower", "randomfollower2");
    }

    @Test
    @SqlGroup({@Sql(value = "classpath:populatedb.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)})
    public void testGetFollowers() {
        given(usersClientService.userExists(anyString())).willReturn(true);

        final ResponseEntity<FollowersResponse> resp = restTemplate.getForEntity(url("/api/v1/connections/johndoe"), FollowersResponse.class);
        assertThat(resp.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(resp.getBody().getFollowers().size()).isEqualTo(2);
        assertThat(resp.getBody().getFollowers()).contains("jojobizarre", "laracroft");

        final ResponseEntity<FollowersResponse> resp2 = restTemplate.getForEntity(url("/api/v1/connections/lonelyguy"), FollowersResponse.class);
        assertThat(resp2.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(resp2.getBody().getFollowers().size()).isEqualTo(1);
        assertThat(resp2.getBody().getFollowers()).contains("lonelyguy");

        final ResponseEntity<FollowingsResponse> resp3 = restTemplate.getForEntity(url("/api/v1/connections/laracroft/followings"), FollowingsResponse.class);
        assertThat(resp3.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(resp3.getBody().getFollowings().size()).isEqualTo(1);
        assertThat(resp3.getBody().getFollowings()).contains("johndoe");
    }

    @Test
    @SqlGroup({@Sql(value = "classpath:populatedb.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)})
    public void testAddConnection() {
        given(usersClientService.userExists(anyString())).willReturn(true);

        final ResponseEntity<FollowersResponse> resp = restTemplate.postForEntity(url("/api/v1/connections/lonelyguy/shadowman"), null, FollowersResponse.class);
        assertThat(resp.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(resp.getBody().getFollowers().size()).isEqualTo(2);
        assertThat(resp.getBody().getFollowers()).contains("lonelyguy", "shadowman");
    }

    @Test
    public void testAddConnectionUnknownUser() {
        given(usersClientService.userExists("randomuser")).willReturn(false);
        given(usersClientService.userExists("randomfollower")).willReturn(true);

        final ResponseEntity<FollowersResponse> resp = restTemplate.postForEntity(url("/api/v1/connections/randomuser/randomfollower"), null, FollowersResponse.class);
        assertThat(resp.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);
    }

    @Test
    public void testAddConnectionUnknownFollower() {
        given(usersClientService.userExists("randomuser")).willReturn(true);
        given(usersClientService.userExists("randomfollower")).willReturn(false);

        final ResponseEntity<FollowersResponse> resp = restTemplate.postForEntity(url("/api/v1/connections/randomuser/randomfollower"), null, FollowersResponse.class);
        assertThat(resp.getStatusCode()).isEqualTo(HttpStatus.NOT_FOUND);
    }

    @Test
    @SqlGroup({@Sql(value = "classpath:populatedb.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)})
    public void testRemoveFollowers() {
        given(usersClientService.userExists(anyString())).willReturn(true);

        restTemplate.delete(url("/api/v1/connections/johndoe/laracroft"));
        final ResponseEntity<FollowersResponse> resp = restTemplate.getForEntity(url("/api/v1/connections/johndoe"), FollowersResponse.class);
        assertThat(resp.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(resp.getBody().getFollowers()).containsExactlyInAnyOrder("jojobizarre");
    }
}

File: yatc-posts/src/test/kotlin/fr/alexandreroman/yatc/posts/ApplicationTests.kt
/*
 * Copyright (c) 2019 Pivotal Software, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.alexandreroman.yatc.posts

import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.BDDMockito.given
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.cloud.stream.messaging.Source
import org.springframework.cloud.stream.test.binder.MessageCollector
import org.springframework.http.HttpStatus
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit4.SpringRunner

@ActiveProfiles("test")
@RunWith(SpringRunner::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class ApplicationTests {
    @LocalServerPort
    private var webPort = 0
    @Autowired
    private lateinit var restTemplate: TestRestTemplate

    @MockBean
    private lateinit var usersClientService: UsersClientService
    @Autowired
    private lateinit var messageCollector: MessageCollector
    @Autowired
    private lateinit var source: Source

    private fun url(path: String) = "http://localhost:$webPort$path"

    @Test
    fun contextLoads() {
    }

    @Test
    fun testCreatePostUnknownAuthor() {
        given(usersClientService.userExists(anyString())).willReturn(false)

        val req = NewPostRequest("johndoe", "Hello")
        val resp = restTemplate.postForEntity(url("/api/v1/posts"), req, String::class.java)
        assertThat(resp.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
    }

    @Test
    fun testCreatePost() {
        given(usersClientService.userExists(anyString())).willReturn(true)
        given(usersClientService.getUser("joe")).willReturn(User("joe"))

        val req = NewPostRequest("joe", "Hello")
        val resp = restTemplate.postForEntity(url("/api/v1/posts"), req, Post::class.java)
        assertThat(resp.statusCode).isEqualTo(HttpStatus.OK)
        val post = resp.body!!
        assertThat(post.content).isEqualTo("Hello")
        assertThat(post.created).isNotNull()
        assertThat(post.id).isEqualTo(1)

        val msg = messageCollector.forChannel(source.output()).take()
        val om = jacksonObjectMapper().registerModule(JavaTimeModule())
        val receivedPost = om.readValue(msg.payload.toString(), Post::class.java)
        assertThat(receivedPost.id).isEqualTo(1)
        assertThat(receivedPost.author).isEqualTo("joe")
        assertThat(receivedPost.created).isNotNull()
        assertThat(receivedPost.content).isEqualTo("Hello")

        val resp2 = restTemplate.getForEntity(url("/api/v1/posts/1"), Post::class.java)
        assertThat(resp2.statusCode).isEqualTo(HttpStatus.OK)
        val post2 = resp2.body!!
        assertThat(post2).isEqualTo(post)
    }

    @Test
    fun testGetPostUnknown() {
        val resp = restTemplate.getForEntity(url("/api/v1/posts/123"), String::class.java)
        assertThat(resp.statusCode).isEqualTo(HttpStatus.NOT_FOUND)
    }
}
